python/smqtk/algorithms/nn_index/mrpt.py

```diff
@@ -0,0 +1,451 @@
+
+    def count(self):
+        """
+        :return: Number of elements in this index.
+        :rtype: int
+        """
+        super(MRPTNearestNeighborsIndex, self).count()
+        return len(self._descr_cache) if self._descr_cache else 0
+
+    def nn(self, d, n=1):
+        """
+        Return the nearest `N` neighbors to the given descriptor element.
+
+        :param d: Descriptor element to compute the neighbors of.
+        :type d: smqtk.representation.DescriptorElement
+
+        :param n: Number of nearest neighbors to find.
+        :type n: int
+
+        :return: Tuple of nearest N DescriptorElement instances, and a tuple of
+            the distance values to those neighbors.
+        :rtype: (tuple[smqtk.representation.DescriptorElement], tuple[float])
+
+        """
+        super(MRPTNearestNeighborsIndex, self).nn(d, n)
+
+        def _query_single(tree):
+            random_basis = tree['random_basis']
+            depth = random_basis.shape[1]
+            proj_query = d.vector() * random_basis
+            splits = tree['splits']
+            idx = 0
+            for level in range(depth):
+                split_point = splits[idx]
+                # Look at the level'th coordinate of proj_query
+                if proj_query[level] < split_point:
+                    idx = 2 * idx + 1
+                else:
+                    idx = 2 * idx + 2
+
+            # idx will be `2^depth - 1` greater than the position of the leaf
+            # in the list
+            idx -= ((1 << depth) - 1)
+            return tree['leaves'][idx]
+
+        def _exact_query(idcs):
+            # Assemble the array to query from the descriptors that match
+            pts_array = [self._descr_cache[idx].vector() for idx in idcs]
+            pts_array = np.array(pts_array, dtype=pts_array[0].dtype)
+
+            dists = ((pts_array - d.vector()) ** 2).sum(axis=1)
+
+            if n >= dists.shape[0]:
+                return idcs, dists
+
+            near_indices = np.argpartition(dists, n - 1)[:n]
+            return idcs[near_indices], dists[near_indices]
+
+        # Use the sparse matrix trick to count the votes:
+        # I.e. construct a sparse matrix with ones as data, one row, and the
+        # indices as columns. After we add the votes for every tree, we extract
+        # the indices with enough votes.
+        votes = csr_matrix((1, len(self._descr_cache)), dtype=np.int32)
+
+        for t in self._trees:
+            leaf = _query_single(t)
+
+            # Use the (data, indices, indptr) constructor. For one row, it's
+            # simple
+            votes += csr_matrix(
+                (np.ones_like(leaf), leaf, [0, leaf.size]),
+                shape=votes.shape, dtype=votes.dtype)
+
+        # Votes will be in CSR format after addition. We get indices for free
+        # after computing the comparison.
+        sufficient_votes = votes.__ge__(self._required_votes)
```

Comment:
I think it was that PyCharm quirk where it gets convinced that every relational operator returns bool, so I was verifying that the types all lined up, and left it in. Will clean.